### ISSUE TITLE
NID_safeContentsBag wants SET-OF, RFC says SEQUENCE-OF

### ISSUE THREAD

**Issue by bartonjs** [openssl/openssl#6665]
The ASN.1 module in IETF RFC 7292 (https://tools.ietf.org/html/rfc7292#appendix-D) says

```asn.1
SafeContents ::= SEQUENCE OF SafeBag

safeContentsBag BAG-TYPE ::=
     {SafeContents        IDENTIFIED BY {bagtypes 6}}
```

So a SafeContentsBag value should look like

```
SEQUENCE
    OID("1.2.840.113549.1.12.10.1.6"),
    CONSTRUCTED [0]
        SEQUENCE(-OF)
            bag0, ...
```

The ASN in [RSA PKCS#12](https://www.emc.com/collateral/white-papers/h11301-pkcs-12v1-1-personal-information-exchange-syntax-wp.pdf) (2012) is the same, and I haven't managed to find a copy of the 1997 version.

But the implementation ASN says it's a SET-OF, not a SEQUENCE-OF.

https://github.com/openssl/openssl/blob/6f0ac0e2f27d9240516edb9a23b7863e7ad02898/crypto/pkcs12/p12_asn.c#L51-L58

It looks like the original implementation had writing as SEQUENCE-OF and reading as SET-OF, and the SET-OF is what got maintained during the ASN.1 template conversion:

https://github.com/openssl/openssl/blob/8d8c7266d4de9887fb0190a0770df9dc254a16a3/crypto/pkcs12/p12_sbag.c#L91-L94

https://github.com/openssl/openssl/blob/8d8c7266d4de9887fb0190a0770df9dc254a16a3/crypto/pkcs12/p12_sbag.c#L127-L130

https://github.com/openssl/openssl/blob/8d8c7266d4de9887fb0190a0770df9dc254a16a3/crypto/pkcs12/p12_sbag.c#L187-L191

**Comment by richsalz** [comment]
So this has been a bug for 20 years?

**Comment by bartonjs** [comment]
Yeah, as far as I can tell.

Given that Windows PFXImportCertStore doesn't seems to read SafeContentsBags I don't understand why they ever made it into the spec.  OpenSSL is the only PFX/PKCS12 implementation that I saw which even nominally bothered with SafeContentsBag, and I'd believe that (almost) no one has ever bothered making use of one.

Certainly a few obvious options present themselves:
* No one has noticed / cared, leave it as-is.
* Probably no one uses it, change it to match the spec.
* Read via a CHOICE, write it as SEQUENCE-OF (spec-compliant).
* Read via a CHOICE, write it as SET-OF (backwards-compatible).

FWIW, I discovered it because I was adding a PKCS12 object API to .NET, and found that `openssl pkcs12` couldn't read my file.  Then I found that none of the other libraries that I could think of (or find) for PKCS12 bothered with it (though many had a version of "// TODO: handle this").

**Comment by levitte** [comment]
Did you try changing it to a SEQUENCE-OF in your OpenSSL source copy?  Did it work correctly with your .NET API?  Both directions?

If no one else bothers with it (which does seem likely, or we would probably have heard of this much earlier), I see absolutely no harm in simply changing to match the spec.

**Comment by davidben** [comment]
Skimming the NSS implementation, they seem to do something with it and use SEQUENCE, per the spec.

**Comment by levitte** [comment]
That does it

**Comment by bartonjs** [comment]
Here's a test output from my beta API with a local change to 1.0.2g (where I happened to be synced).

```
$ git status
HEAD detached at OpenSSL_1_0_2g
```

```diff
--- a/crypto/pkcs12/p12_asn.c
+++ b/crypto/pkcs12/p12_asn.c
@@ -100,7 +100,7 @@ ASN1_ADB_TEMPLATE(safebag_default) = ASN1_EXP(PKCS12_SAFEBAG, value.other, ASN1_
 ASN1_ADB(PKCS12_SAFEBAG) = {
         ADB_ENTRY(NID_keyBag, ASN1_EXP(PKCS12_SAFEBAG, value.keybag, PKCS8_PRIV_KEY_INFO, 0)),
         ADB_ENTRY(NID_pkcs8ShroudedKeyBag, ASN1_EXP(PKCS12_SAFEBAG, value.shkeybag, X509_SIG, 0)),
-        ADB_ENTRY(NID_safeContentsBag, ASN1_EXP_SET_OF(PKCS12_SAFEBAG, value.safes, PKCS12_SAFEBAG, 0)),
+        ADB_ENTRY(NID_safeContentsBag, ASN1_EXP_SEQUENCE_OF(PKCS12_SAFEBAG, value.safes, PKCS12_SAFEBAG, 0)),
         ADB_ENTRY(NID_certBag, ASN1_EXP(PKCS12_SAFEBAG, value.bag, PKCS12_BAGS, 0)),
         ADB_ENTRY(NID_crlBag, ASN1_EXP(PKCS12_SAFEBAG, value.bag, PKCS12_BAGS, 0)),
         ADB_ENTRY(NID_secretBag, ASN1_EXP(PKCS12_SAFEBAG, value.bag, PKCS12_BAGS, 0))
```


Data file is

* PFX
  * AuthSafe
    * Unencrypted safeContents
      * SafeContentsBag
        * CertBag (LocalKeyId { 01 })
      * CertBag (LocalKeyId { 02 })
  * MACData (password: "hi", PRF=SHA1, IterationCount=2048)

```
$ xxd -p -r > nested.pfx
30820514020103308204D006092A864886F70D010701A08204C1048204BD308204B9308202810
6092A864886F70D010701A08202720482026E3082026A30820266060B2A864886F70D010C0A01
06A0820255308202513082024D060B2A864886F70D010C0A0103A082022830820224060A2A864
886F70D01091601A0820214048202103082020C30820179A00302010202105D2FFFF863BABC9B
4D3C80AB178A4CCA300906052B0E03021D0500301E311C301A060355040313135253414B65795
472616E736665724361706931301E170D3135303431353037303030305A170D32353034313530
37303030305A301E311C301A060355040313135253414B65795472616E7366657243617069313
0819F300D06092A864886F70D010101050003818D0030818902818100AA272700586C0CC41B05
C65C7D846F5A2BC27B03E301C37D9BFF6D75B6EB6671BA9596C5C63BA2B1AF5C318D9CA39E740
0D10C238AC72630579211B86570D1A1D44EC86AA8F6C9D2B4E283EA3535923F398A312A23EAEA
CD8D34FAACA965CD910B37DA4093EF76C13B337C1AFAB7D1D07E317B41A336BAA4111299F9942
4408D0203010001A3533051304F0603551D0104483046801015432DB116B35D07E4BA89EDB246
9D7AA120301E311C301A060355040313135253414B65795472616E73666572436170693182105
D2FFFF863BABC9B4D3C80AB178A4CCA300906052B0E03021D05000381810081E5535D8ECEEF26
5ACBC82F6C5F8BC9D84319265F3CCF23369FA533C8DC1938952C5931662D9ECD8B1E7B81749E4
8468167E2FCE3D019FA70D54646975B6DC2A3BA72D5A5274C1866DA6D7A5DF47938E034A075D1
1957D653B5C78E5291E4401045576F6D4EDA81BEF3C369AF56121E49A083C8D1ADB09F291822E
99A4296463112301006092A864886F70D01091531030401013082023006092A864886F70D0107
01A08202210482021D3082021930820215060B2A864886F70D010C0A0103A08201F0308201EC0
60A2A864886F70D01091601A08201DC048201D8308201D43082013DA003020102021072C6C773
4916468C4D608253DA017676300D06092A864886F70D01010B05003020311E301C06035504031
3155253415368613235364B65795472616E7366657231301E170D313630343138313035393436
5A170D3137303431383136353934365A3020311E301C060355040313155253415368613235364
B65795472616E736665723130819F300D06092A864886F70D010101050003818D003081890281
8100CAD046DE3A7F6DC78FC5A4E01D1F7D90DB596F586334D5708AECB8E52D6BB912C0B5EC963
3A82B4ABAC4C2860C766F2FDF1C905C4A72A54ADFD041ADABE5F2AFD1E2AD88615970E818DC3D
4D00BB6C4CE94C5EB4E3EFEDD80D14C3D295EA471AE430CBB20B071582F1396369FBE90C14AA5
F85B8E3B14011D81FBD41ECB1495D0203010001A30F300D300B0603551D0F040403020520300D
06092A864886F70D01010B050003818100BAED2A5AE2D11EE4209C0694C790E72E3E8AD310B25
06B277D7C001B09F660D48DBA846AC5BBEF97653613ADF53D7624FC9B2B337F25CB3374227900
CFEFBE2FDAC92B4F769CF2BF3BEFB485F282A85BFB09454B797CE5286DE560C219FB0DD6FCE04
42ADBFEF4F767E9AC81CF3E9701BAF81EFC73A0ED88576ADFF12413B8273112301006092A8648
86F70D0109153103040102303B301F300706052B0E03021A041429011F38FA41F13B045DAEB42
264B35D5AFCC7A80414C33491C1681CF43F123CCED392E85E2F7D4FC22502020800
```

Stable library (fails to read):

```
$ openssl pkcs12 -in nested.pfx -info -nodes -passin pass:hi
MAC Iteration 2048
MAC verified OK
PKCS7 Data
Error outputting keys and certificates
140128844408472:error:0D0680A8:asn1 encoding routines:ASN1_CHECK_TLEN:wrong tag:tasn_dec.c:1217:
140128844408472:error:0D08303A:asn1 encoding routines:ASN1_TEMPLATE_NOEXP_D2I:nested asn1 error:tasn_dec.c:647:
140128844408472:error:0D08403A:asn1 encoding routines:ASN1_TEMPLATE_EX_D2I:nested asn1 error:tasn_dec.c:580:Field=value.safes, Type=PKCS12_SAFEBAG
140128844408472:error:0D08303A:asn1 encoding routines:ASN1_TEMPLATE_NOEXP_D2I:nested asn1 error:tasn_dec.c:689:
140128844408472:error:0D0C706E:asn1 encoding routines:ASN1_item_unpack:decode error:asn_pack.c:205:
```

With the edit:

```
$ LD_PRELOAD=/home/jbarton/git/openssl/openssl/libcrypto.so.1.0.0 openssl pkcs12 -in nested.pfx -info -nodes -passin pass:hi
openssl: /home/jbarton/git/openssl/openssl/libcrypto.so.1.0.0: no version information available (required by openssl)
openssl: /home/jbarton/git/openssl/openssl/libcrypto.so.1.0.0: no version information available (required by openssl)
openssl: /home/jbarton/git/openssl/openssl/libcrypto.so.1.0.0: no version information available (required by openssl)
openssl: /home/jbarton/git/openssl/openssl/libcrypto.so.1.0.0: no version information available (required by openssl)
openssl: /home/jbarton/git/openssl/openssl/libcrypto.so.1.0.0: no version information available (required by /lib/x86_64-linux-gnu/libssl.so.1.0.0)
openssl: /home/jbarton/git/openssl/openssl/libcrypto.so.1.0.0: no version information available (required by /lib/x86_64-linux-gnu/libssl.so.1.0.0)
openssl: /home/jbarton/git/openssl/openssl/libcrypto.so.1.0.0: no version information available (required by /lib/x86_64-linux-gnu/libssl.so.1.0.0)
openssl: /home/jbarton/git/openssl/openssl/libcrypto.so.1.0.0: no version information available (required by /lib/x86_64-linux-gnu/libssl.so.1.0.0)
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
MAC Iteration 2048
MAC verified OK
PKCS7 Data
Safe Contents bag
Bag Attributes: <No Attributes>
Certificate bag
Bag Attributes
    localKeyID: 01
subject=/CN=RSAKeyTransferCapi1
issuer=/CN=RSAKeyTransferCapi1
-----BEGIN CERTIFICATE-----
MIICDDCCAXmgAwIBAgIQXS//+GO6vJtNPICrF4pMyjAJBgUrDgMCHQUAMB4xHDAa
BgNVBAMTE1JTQUtleVRyYW5zZmVyQ2FwaTEwHhcNMTUwNDE1MDcwMDAwWhcNMjUw
NDE1MDcwMDAwWjAeMRwwGgYDVQQDExNSU0FLZXlUcmFuc2ZlckNhcGkxMIGfMA0G
CSqGSIb3DQEBAQUAA4GNADCBiQKBgQCqJycAWGwMxBsFxlx9hG9aK8J7A+MBw32b
/211tutmcbqVlsXGO6Kxr1wxjZyjnnQA0QwjiscmMFeSEbhlcNGh1E7Iaqj2ydK0
4oPqNTWSPzmKMSoj6urNjTT6rKllzZELN9pAk+92wTszfBr6t9HQfjF7QaM2uqQR
Epn5lCRAjQIDAQABo1MwUTBPBgNVHQEESDBGgBAVQy2xFrNdB+S6ie2yRp16oSAw
HjEcMBoGA1UEAxMTUlNBS2V5VHJhbnNmZXJDYXBpMYIQXS//+GO6vJtNPICrF4pM
yjAJBgUrDgMCHQUAA4GBAIHlU12Ozu8mWsvIL2xfi8nYQxkmXzzPIzafpTPI3Bk4
lSxZMWYtns2LHnuBdJ5IRoFn4vzj0Bn6cNVGRpdbbcKjunLVpSdMGGbabXpd9Hk4
4DSgddEZV9ZTtceOUpHkQBBFV29tTtqBvvPDaa9WEh5JoIPI0a2wnykYIumaQpZG
-----END CERTIFICATE-----
PKCS7 Data
Certificate bag
Bag Attributes
    localKeyID: 02
subject=/CN=RSASha256KeyTransfer1
issuer=/CN=RSASha256KeyTransfer1
-----BEGIN CERTIFICATE-----
MIIB1DCCAT2gAwIBAgIQcsbHc0kWRoxNYIJT2gF2djANBgkqhkiG9w0BAQsFADAg
MR4wHAYDVQQDExVSU0FTaGEyNTZLZXlUcmFuc2ZlcjEwHhcNMTYwNDE4MTA1OTQ2
WhcNMTcwNDE4MTY1OTQ2WjAgMR4wHAYDVQQDExVSU0FTaGEyNTZLZXlUcmFuc2Zl
cjEwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAMrQRt46f23Hj8Wk4B0ffZDb
WW9YYzTVcIrsuOUta7kSwLXsljOoK0q6xMKGDHZvL98ckFxKcqVK39BBravl8q/R
4q2IYVlw6BjcPU0Au2xM6UxetOPv7dgNFMPSlepHGuQwy7ILBxWC8TljafvpDBSq
X4W447FAEdgfvUHssUldAgMBAAGjDzANMAsGA1UdDwQEAwIFIDANBgkqhkiG9w0B
AQsFAAOBgQC67Spa4tEe5CCcBpTHkOcuPorTELJQayd9fAAbCfZg1I26hGrFu++X
ZTYTrfU9diT8myszfyXLM3QieQDP774v2skrT3ac8r8777SF8oKoW/sJRUt5fOUo
beVgwhn7Ddb84EQq2/7092fprIHPPpcBuvge/HOg7YhXat/xJBO4Jw==
-----END CERTIFICATE-----
```

**Comment by levitte** [comment]
Cool. Fix présent in #6708

**Comment by t-j-h** [comment]
FYI - the official version 1.0 of PKCS12 (the 1999) version also lists it correctly as SEQUENCE OF as SET OF would not have made sense
